### PR TITLE
[TOPIC-BLE-LLCP] Bluetooth: controller: Implementing support for TX ctrl buffer resource management

### PR DESCRIPTION
**Pull request by erbr-ot**
**Control Tx Buffer alloc considerations**

With no resource management the worst case need for tx ctrl buffers for a given connection is 4:
an ongoing encryption procedure plus a possible rejection to a remote request as well as a local terminate
So, if there are MAX_CONN x 4 ctrl tx buffers available for LLCP no resource management is needed

With a queueing system however, the system worst case need could be as low as 1 tx buffer in total.
With 1 tx buffer in total for LLCP, a conn/proc cannot however be guarantied a buffer as it could be in use (and possibly stuck)
'in' another conn/proc. This seems to be a poor design for most cases.

With 1 buffer per connection on the other hand, it can be guarantied that a conn/proc will not stall on acount
of other connections, however, if there is only one per connection, and with the LLCP 'engine' tied to prepare/done
the maximum throughput tx ctrl wise would be 1 PDU per connection event.

For the encryption procedure this '1 PDU per event' limitation would mean that enc_req/enc_start PDUs on a peripheral
and the enc_pause/enc_start PDUs on the central would be in subsequent conn. events and not the same conn. event
The same would be the case for a remote vs local procedures, that only one tx PDU is possible per event, thus a
procedure request (local or remote) would not nescesarily result in a tx PDU in the following conn. event,
but could be delayed one event (or more, depending on the queueing paradigm)

If a tx ctrl PDU gets 'stuck', because a connection needs re-tx'ing this has little impact re. tx buffer alloc
as the connection is in a limbo anyhow as long as a PDU is not ack'ed.

Having a tx ctrl buffer alloc mechanism allows having a per connection pre-alloc in combination with a cross conn/proc pool of 'extra'
buffers. This could mean that in 'most cases' a procedure could allocate more than one tx buffer while still
allowing the controller to be able to respond to remote request without delay, thus ensuring that any procedure on any connection
can be handled at least on a per connection event basis.

**Paradigm presented in this PR**
By KConfig a certain number of tx buffers per connection plus a certain number of common tx buffers are declared, and allocation of these are handled through a wait list. The concept being that each connection always has access to at least the configured number of tx buffers, and can access the common pool on a first come first serve basis

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h [hunk 1]
```diff
@@ -113,6 +118,15 @@ struct proc_ctx {
 	/* Procedure pause */
 	int pause;
 
+	/* Handle of related connection */
+	uint16_t conn_handle;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:122]: As we discussed Thursday, remove the use of conn_handle and use the conn pointer instead.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h [hunk 2]
```diff
@@ -28,7 +28,7 @@ void ull_cp_state_set(struct ll_conn *conn, uint8_t state);
 /**
 *
 */
-void ull_cp_release_tx(struct node_tx *tx);
+void ull_cp_release_tx(uint16_t handle, struct node_tx *tx);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.h:31]: Use a `struct ll_conn*` instead

File: subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h [hunk 3]
```diff
@@ -284,13 +304,15 @@ static inline bool is_instant_reached_or_passed(uint16_t instant, uint16_t event
...
 bool llcp_ntf_alloc_is_available(void);
 bool llcp_ntf_alloc_num_available(uint8_t count);
 struct node_rx_pdu *llcp_ntf_alloc(void);
-struct proc_ctx *llcp_create_local_procedure(enum llcp_proc proc);
-struct proc_ctx *llcp_create_remote_procedure(enum llcp_proc proc);
+struct proc_ctx *llcp_create_local_procedure(uint16_t conn_handle, enum llcp_proc proc);
+struct proc_ctx *llcp_create_remote_procedure(uint16_t conn_handle, enum llcp_proc proc);
+bool llcp_tx_alloc_peek(struct proc_ctx *ctx);
+void llcp_tx_alloc_unpeek(struct proc_ctx *ctx);
+ctx_handle_t llcp_proc_ctx_handle_get(struct proc_ctx *ctx);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:314]: Is this still necessary?

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h (hunk 3 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:314]: no

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h (hunk 2 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.h:31]: yeap

File: subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h (hunk 1 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:122]: yeap

**Comment by erbr-ot** [comment]
Rebased and added a couple #ifdef's

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 4]
```diff
@@ -90,18 +84,115 @@ void llcp_proc_ctx_release(struct proc_ctx *ctx)
...
+	return true;
+}
+
+/*
+ * @brief un-peek of a tx buffer, in case ongoing alloc is aborted
+ *
+ */
+void llcp_tx_alloc_unpeek(struct proc_ctx *ctx)
+{
+	sys_slist_remove(&tx_buffer_wait_list, NULL, &ctx->wait_node);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:142]: Can we always be sure that unpeek is done on a head node as the NULL indicates? Would `sys_slist_find_and_remove()` be better?

Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c:142]: No, we cannot - I had understood the remove as doing find and remove, when passing NULL as previous node. I will replace!

File: subsys/bluetooth/controller/Kconfig.ll_sw_split [hunk 5]
```diff
@@ -453,6 +453,47 @@ config BT_CTLR_LLCP_CONN
...
+	  This defines the size of the pool of control procedure contexts available
+	  for handlign control procedures. This pool is shared across all
+	  connections (local vs remote initiate), with allocation through a queue
+
+config BT_CTLR_LLCP_TX_BUFFER_QUEUE_SIZE
+	int "Number of entries in control procedure contexts tx buffer request queue"
+	default BT_CTLR_LLCP_PROC_CTX_BUF_NUM
+	range 0 BT_CTLR_LLCP_PROC_CTX_BUF_NUM
+	help
+	  Set the size of the tx buffer request queue.
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/Kconfig.ll_sw_split:492]: This doesn't seem to make sense anymore. Should this be a bool, that would be calculated based on the value of BT_CTLR_LLCP_PER_CONN_TX_CTRL_BUF_NUM?

File: subsys/bluetooth/controller/Kconfig.ll_sw_split [hunk 6]
```diff
@@ -453,6 +453,47 @@ config BT_CTLR_LLCP_CONN
...
+
+config BT_CTLR_LLCP_COMMON_TX_CTRL_BUF_NUM
+	int "Number of tx control buffers to be available across all connections"
+	default BT_MAX_CONN
+	range 1 255
+	help
+	  Set the number control buffers that is to be available for tx.
+	  This defines the size of the pool of tx buffers available
+	  for control procedure tx. This pool is shared across all
+	  procedures/connections with allocation through a fifo queue
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/Kconfig.ll_sw_split:475]: If every connection has it's theoretical maximum number of buffers, then there is no need for a shared common pool of buffers, correct?

File: tests/bluetooth/controller/common/src/helper_util.c [hunk 7]
```diff
@@ -190,10 +191,13 @@ void test_print_conn(struct ll_conn *conn)
 	printf("     Reject: %d\n", conn->llcp.remote.reject_opcode);
 	printf("--------------------->\n");
 }
-void test_setup(struct ll_conn *conn)
+void test_setup(struct ll_conn **conn)
```
Comment (thoh-ot) [inline tests/bluetooth/controller/common/src/helper_util.c]: There should no longer be a need to do this as the buffer allocation freeeing no longer needs a conn handle and therefor `ll_conn_acquire()` below is no longer needed.

Comment (erbr-ot) [reply tests/bluetooth/controller/common/src/helper_util.c]: Well, not really - the release still needs the conn object to keep track

File: subsys/bluetooth/controller/Kconfig.ll_sw_split (hunk 5 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/Kconfig.ll_sw_split:492]: Yeah, I guess you are right. So if BT_CTLR_LLCP_PER_CONN_TX_CTRL_BUF_NUM < 4?

File: subsys/bluetooth/controller/Kconfig.ll_sw_split (hunk 6 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/Kconfig.ll_sw_split:475]: Correct, I'll change the range to include 0.
Also I was in doubt as to what to make the default? BT_MAX_CONN is probably not so good, as it implies a relation, and there is none!

**Review by thoh-ot — APPROVED** [review]

**Review by kruithofa — CHANGES_REQUESTED** [review]

**Review by kruithofa — CHANGES_REQUESTED** [review]

**Review by kruithofa — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Pull request merged**